File: examples/ts/payments/gateway.ts
// @context decision #examples/docs/context/gate-42.md !critical -- Strict > avoids duplicate processing.
export function shouldProcess(messageTimestamp: number, cutoff: number): boolean {
  return messageTimestamp > cutoff;
}

// @context decision:tradeoff #examples/docs/context/cache-strategy.ctx.md !high -- LRU keeps eviction O(1).
export class LRUCache<K, V> {
  readonly #store = new Map<K, V>();

  get(key: K): V | undefined {
    const value = this.#store.get(key);
    if (value === undefined) {
      return undefined;
    }

    this.#store.delete(key);
    this.#store.set(key, value);
    return value;
  }

  set(key: K, value: V): void {
    this.#store.set(key, value);
  }
}

// @context requirement #examples/docs/requirements/billing-rounding.md !high -- Settlement must match finance rules.
export function roundForSettlement(value: number): number {
  return Math.round(value * 100) / 100;
}

// @context risk:security !critical -- Caller must authenticate the webhook before invoking this parser.
export function parseWebhook(body: string): Record<string, unknown> {
  return JSON.parse(body) as Record<string, unknown>;
}

// @context related #examples/go/payments/gateway.go -- Keep TS and Go gateway behavior aligned.
export function nextRetryCount(base: number): number {
  return base + 1;
}

// @context history -- Replaced recursive retry scheduling after stack growth in production.
export function scheduleRetries(maxRetries: number): number[] {
  const retries: number[] = [];
  for (let attempt = 1; attempt <= maxRetries; attempt += 1) {
    retries.push(attempt);
  }
  return retries;
}

// @context doc -- This helper keeps call sites focused on business logic rather than HTTP boilerplate.
export async function postJSON(url: string, payload: unknown): Promise<Response> {
  return fetch(url, {
    method: "POST",
    headers: { "content-type": "application/json" },
    body: JSON.stringify(payload),
  });
}

// @context:decision:assumption -- Legacy syntax still parses; callers assume p99 < 500ms.
export const maxRetries = 3;

File: examples/go/payments/gateway.go
package payments

import "encoding/json"

// @context decision #examples/docs/context/gate-42.md !critical [verified:2026-03-24] -- Strict > avoids duplicate processing.
func ShouldProcess(messageTimestamp int64, cutoff int64) bool {
	return messageTimestamp > cutoff
}

// @context decision:constraint !high [verified:2026-03-24] -- Upstream caps batch size at 100, so keep this guard local.
func ClampBatchSize(size int) int {
	if size > 100 {
		return 100
	}
	return size
}

// @context decision:assumption [verified:2026-03-24] -- Retry budget assumes upstream p99 latency stays under 500ms.
const MaxRetries = 3

// @context requirement #examples/docs/requirements/billing-rounding.md [verified:2026-03-24] -- Finance requires stable settlement rounding.
func RoundForSettlement(cents int64) int64 {
	return cents
}

// @context risk:perf !high [verified:2026-03-24] -- Re-encoding on every call is acceptable only because payloads stay small.
func EncodeEnvelope(payload map[string]any) ([]byte, error) {
	return json.Marshal(payload)
}

// @context risk:security !critical [verified:2026-03-24] -- Caller must validate the webhook signature before decoding.
func DecodeWebhook(body []byte) (map[string]any, error) {
	decoded := map[string]any{}
	err := json.Unmarshal(body, &decoded)
	return decoded, err
}

// @context related #examples/ts/payments/gateway.ts [verified:2026-03-24] -- Keep Go and TS gateway semantics aligned.
func NextRetryCount(base int) int {
	return base + 1
}

// @context history [verified:2026-03-24] -- This loop replaced recursive retry scheduling after stack growth in production.
func RetrySchedule(maxRetries int) []int {
	retries := make([]int, 0, maxRetries)
	for attempt := 1; attempt <= maxRetries; attempt++ {
		retries = append(retries, attempt)
	}
	return retries
}

// @context doc [verified:2026-03-24] -- This helper centralizes JSON encoding so call sites stay narrow.
func MustEncode(payload map[string]any) []byte {
	body, _ := json.Marshal(payload)
	return body
}

// @context:decision:tradeoff #examples/docs/context/cache-strategy.ctx.md -- Legacy form remains accepted during migration.
func CacheStrategy() string {
	return "lru"
}
